File: packages/ts/core/src/pagination.ts
import type {
  ResponseMeta,
  SuccessEnvelope,
} from "./types.js";

export async function* paginate<TItem, TMeta = ResponseMeta>(
  fetchPage: (cursor?: string) => Promise<SuccessEnvelope<TItem[], TMeta>>,
): AsyncGenerator<TItem, void, void> {
  let cursor: string | undefined;

  do {
    const page = await fetchPage(cursor);
    for (const item of page.data) {
      yield item;
    }

    if (page.pagination?.has_more) {
      if (!page.pagination.next_cursor) {
        throw new Error("Sendmux pagination response had has_more=true without next_cursor");
      }

      cursor = page.pagination.next_cursor;
    } else {
      cursor = undefined;
    }
  } while (cursor);
}

File: packages/python/core/sendmux_core/pagination.py
from __future__ import annotations

from collections.abc import Callable, Iterator, Sequence
from dataclasses import dataclass
from typing import Generic, Protocol, TypeVar

T = TypeVar("T")


class PaginationMeta(Protocol):
    has_more: bool
    next_cursor: str | None


class CursorResponse(Protocol[T]):
    data: Sequence[T]
    pagination: PaginationMeta


@dataclass(frozen=True)
class CursorPage(Generic[T]):
    data: Sequence[T]
    has_more: bool
    next_cursor: str | None


def iter_cursor_pages(fetch_page: Callable[[str | None], CursorResponse[T]]) -> Iterator[T]:
    cursor: str | None = None

    while True:
        response = fetch_page(cursor)
        for item in response.data:
            yield item

        if not response.pagination.has_more:
            return

        next_cursor = response.pagination.next_cursor
        if not next_cursor or next_cursor == cursor:
            raise RuntimeError("Sendmux cursor pagination did not return a new next_cursor")
        cursor = next_cursor
